### PR TITLE
Provide support for pagination in query operations

### PR DESCRIPTION
**Pull request by sachinira**
# Description
Fixes: https://github.com/ballerina-platform/ballerina-extended-library/issues/377

One line release note: 
- Provide streaming support for the response  of `getQueryResult`  and `searchSOSLString` operations

## Type of change
Please delete options that are not relevant.
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?
Tested for all the available scenarios
Newly added test cases

**Test Configuration**:
* Ballerina Version:  2201.0.2
* Operating System: Ubuntu 20.04 LTS
* Java SDK: 11

# Checklist:
### Security checks
 - [x] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? 
 - [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: salesforce/modules/bulk/client.bal [hunk 1]
```diff
@@ -91,11 +91,11 @@ public isolated client class Client {
...
     @display {label: "Create Job"}
-    isolated remote function createJob(@display {label: "Operation"} Operation operation, 
-                                       @display {label: "SObject Name"} string sobj,
-                                       @display {label: "Content Type"} JobType contentType, 
-                                       @display {label: "External ID Field Name"} string extIdFieldName = "") 
-                                       returns @tainted @display {label: "Bulk Job"} error|BulkJob {
+    isolated remote function createJob(@display {label: "Operation"} Operation operation,
+                                        @display {label: "SObject Name"} string sobj,
+                                        @display {label: "Content Type"} JobType contentType,
+                                        @display {label: "External ID Field Name"} string extIdFieldName = "")
```
Comment (RolandHewage) [inline salesforce/modules/bulk/client.bal:97]: Shall we remove the extra spaces here.

File: salesforce/modules/bulk/client.bal [hunk 2]
```diff
@@ -91,11 +91,11 @@ public isolated client class Client {
...
+                                        returns @display {label: "Bulk Job"} error|BulkJob {
```
Comment (RolandHewage) [inline salesforce/modules/bulk/client.bal:98]: Is the tainted check ignored here from the original code?

File: salesforce/tests/rest_api_test.bal [hunk 3]
```diff
@@ -115,45 +115,54 @@ function testDeleteRecord() {
...
+@test:Config {
+    enable: false
+}
+function testGetQueryResultWithPagination() returns error? {
+    log:printInfo("baseClient -> getQueryResult()");
+    string sampleQuery = "SELECT Name FROM Contact";
+    stream<record {}, error?> resultStream = check baseClient->getQueryResult(sampleQuery);
+    int count = check countStream(resultStream);
+    log:printInfo("Number of records", count = count);
+    test:assertTrue(count > 2000, msg = "Found less than 2000 search records!");
```
Comment (RolandHewage) [inline salesforce/tests/rest_api_test.bal:144]: Suggestion
```
test:assertTrue(count >= 2000, msg = "Found less than 2000 search records!");
```

**Review by RolandHewage — COMMENTED** [review]
Minor formatting changes to re-check. Rest seems ok.

File: salesforce/samples/rest_api_usecases/search_record_by_string.bal [hunk 4]
```diff
@@ -31,17 +31,20 @@ public function main(){
...
-        log:printError(res.message());
-    }
+    stream<record{}, error?> resultStream = check baseClient->searchSOSLString(searchString);
+    int count = check countStream(resultStream);
+    log:printInfo(string `${count} Record Received`);
+}
 
+isolated function countStream(stream<record{}, error?> resultStream) returns int|error{
+    int nLines = 0;
+    var _ = check from var i in resultStream
```
Comment (kasthuriraajan) [inline salesforce/samples/rest_api_usecases/search_record_by_string.bal:43]: Can't we replace `var _` with only  `_`  here?

File: salesforce/stream-implementer.bal [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+isolated function covertSearchResultsToRecordsArray(map<json>[] queryResultArray) returns record {}[]|error {
+    record {}[] resultRecordArray = [];
+    foreach map<json> queryResult in queryResultArray {
+        resultRecordArray.push(check queryResult.cloneWithType(Record));
+    }
+    return resultRecordArray;
+}
+
+type Record record {
```
[the remark this answers is not in the document]
Comment (sachinira) [reply salesforce/stream-implementer.bal:110]: These types are used internally specifically in the stream implementer classes. Therefore, included them in the same file

File: salesforce/modules/bulk/client.bal (hunk 1 above)
Comment (sachinira) [reply salesforce/modules/bulk/client.bal:97]: This was the result from ballerina formatter. Fixing it manually would take a lot of time

File: salesforce/modules/bulk/tests/utils.bal [hunk 6]
```diff
@@ -66,22 +66,20 @@ isolated function checkCsvResult(string result) returns int {
...
-        sfdc:SoqlRecord[]|error records = res.records;
-        if records is sfdc:SoqlRecord[] {
-            string id = records[0]["Id"].toString();
-            contactId = id;
-        } else {
-            test:assertFail(msg = "Getting contact ID by name failed. err=" + records.toString());
-        }
+    string `SELECT Id FROM Contact WHERE FirstName='${firstName}' AND LastName='${lastName}' AND Title='${title}'`;
+    stream<record{}, error?> queryResults = check restClient->getQueryResult(sampleQuery);
+    var result = queryResults.next();
```
[the remark this answers is not in the document]
Comment (sachinira) [reply salesforce/modules/bulk/tests/utils.bal]: Fixed in 8c88af2857fa79a4d714f0fef58d93bab32cb3b9

File: salesforce/tests/rest_api_test.bal (hunk 3 above)
Comment (sachinira) [reply salesforce/tests/rest_api_test.bal]: Fixed in 8c88af2857fa79a4d714f0fef58d93bab32cb3b9

File: salesforce/samples/rest_api_usecases/search_record_by_string.bal (hunk 4 above)
Comment (sachinira) [reply salesforce/samples/rest_api_usecases/search_record_by_string.bal]: We can remove `var _` completely too. Fixed in 8c88af2857fa79a4d714f0fef58d93bab32cb3b9

File: salesforce/samples/bulk_api_usecases/bulk-operations-csv/bulk_delete_csv.bal [hunk 7]
```diff
@@ -110,22 +110,20 @@ public function main() {
...
-        if records is sfdc:SoqlRecord[] {
-            string id = records[0]["Id"].toString();
-            contactId = id;
-        } else {
-            log:printInfo("Getting contact ID by name failed. err=" + records.toString());
-        }
+    string sampleQuery = string `SELECT Id FROM Contact WHERE FirstName='${firstName}' AND LastName='${lastName}' 
+        AND Title='${title}' LIMIT 1`;
+    stream<record {}, error?> queryResults = check baseClient->getQueryResult(sampleQuery);
+    var result = queryResults.next();
```
[the remark this answers is not in the document]
Comment (sachinira) [reply salesforce/samples/bulk_api_usecases/bulk-operations-csv/bulk_delete_csv.bal]: Fixed in 8c88af2857fa79a4d714f0fef58d93bab32cb3b9

File: salesforce/stream-implementer.bal [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+class SoqlQueryResultStream {
```
Comment (LakshanSS) [inline salesforce/stream-implementer.bal:19]:
```suggestion
class SOQLQueryResultStream {
```

File: salesforce/stream-implementer.bal [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+# + attributes - Attribute record
+type SoqlRecord record {|
+    Attribute attributes?;
+    json...;
+|};
+
+# Defines SOSL query result.
+#
+# + searchRecords - Matching records for the given search string
+type SoslSearchResult record {|
```
Comment (LakshanSS) [inline salesforce/stream-implementer.bal:135]:
```suggestion
type SOSLSearchResult record {|
```

File: salesforce/stream-implementer.bal (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasthuriraajan) [reply salesforce/stream-implementer.bal]: ack. +1

File: salesforce/modules/bulk/client.bal (hunk 2 above)
Comment (kasthuriraajan) [reply salesforce/modules/bulk/client.bal:98]: The Taint analyzer is disabled and the `--taint-check` option is removed from SLBeta1. So better we remove IMO. 
Refer the last point here. https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-beta1/#breaking-changes

File: salesforce/stream-implementer.bal (hunk 8 above)
Comment (sachinira) [reply salesforce/stream-implementer.bal]: Fixed in b02c217b06a758477e2c2ae9f981b5960c17658c

File: salesforce/stream-implementer.bal (hunk 9 above)
Comment (sachinira) [reply salesforce/stream-implementer.bal:135]: Fixed in b02c217b06a758477e2c2ae9f981b5960c17658c

**Review by abeykoon — APPROVED** [review]

**Review by LakshanSS — APPROVED** [review]

**Pull request merged**